packages/aws-cdk/test/api/hotswap/hotswap-deployments.test.ts

```diff
@@ -0,0 +1,258 @@
+  expect(mockUpdateLambdaCode).not.toHaveBeenCalled();
+});
+
+test('Resources that are not lambdas or step functions result in a full deployment even when a hotswappable change is present', async () => {
+  // GIVEN
+  setup.setTemplate({
+    Resources: {
+      Func: {
+        Type: 'AWS::Lambda::Function',
+        Properties: {
+          Code: {
+            S3Bucket: 'current-bucket',
+            S3Key: 'current-key',
+          },
+          FunctionName: 'my-function',
+        },
+        Metadata: {
+          'aws:asset:path': 'old-path',
+        },
+      },
+      SomethingElse: {
+        Type: 'AWS::CloudFormation::SomethingElse',
+        Properties: {
+          Prop: 'old-value',
+        },
+      },
+    },
+  });
+  const cdkStackArtifact = setup.cdkStackArtifactOf({
+    template: {
+      Resources: {
+        Func: {
+          Type: 'AWS::Lambda::Function',
+          Properties: {
+            Code: {
+              S3Bucket: 'current-bucket',
+              S3Key: 'new-key',
+            },
+            FunctionName: 'my-function',
+          },
+          Metadata: {
+            'aws:asset:path': 'new-path',
+          },
+        },
+        SomethingElse: {
+          Type: 'AWS::CloudFormation::SomethingElse',
+          Properties: {
+            Prop: 'new-value',
+          },
+        },
+      },
+    },
+  });
+
+  // WHEN
+  const deployStackResult = await setup.tryHotswapDeployment({}, cdkStackArtifact);
+
+  // THEN
+  expect(deployStackResult).toBeUndefined();
+  expect(mockUpdateMachineDefinition).not.toHaveBeenCalled();
+  expect(mockUpdateLambdaCode).not.toHaveBeenCalled();
+});
+
+
+test('changes only to CDK::Metadata result in a noOp', async () => {
+  // GIVEN
+  setup.setTemplate({
+    Resources: {
+      MetaData: {
+        Type: 'AWS::CDK::MetaData',
+        Properties: {
+          Prop: 'old-value',
+        },
+      },
+    },
+  });
+  const cdkStackArtifact = setup.cdkStackArtifactOf({
+    template: {
+      Resources: {
+        MetaData: {
+          Type: 'AWS::CDK::Metadata',
+          Properties: {
+            Prop: 'new-value',
+          },
+        },
+      },
+    },
+  });
+
+  // WHEN
+  const deployStackResult = await setup.tryHotswapDeployment({}, cdkStackArtifact);
+
+  // THEN
+  expect(deployStackResult).not.toBeUndefined();
+  expect(deployStackResult?.noOp).toEqual(true);
+  expect(mockUpdateMachineDefinition).not.toHaveBeenCalled();
+  expect(mockUpdateLambdaCode).not.toHaveBeenCalled();
+});
+
+test('resource deletions require full deployments', async () => {
+  // GIVEN
+  setup.setTemplate({
+    Resources: {
+      Machine: {
+        Type: 'AWS::StepFunctions::StateMachine',
+        Properties: {
+          DefinitionString: {
+            'Fn::Join': [
+              '',
+              [
+                '{ "Prop" : "old-value", "AnotherProp" : "another-old-value" }',
+              ],
+            ],
+          },
```

Comment:
What's the point of the `Fn::Join` here? This can be written equivalently as:

```suggestion
          DefinitionString: '{ "Prop" : "old-value", "AnotherProp" : "another-old-value" }',
```